File: java/src/main/java/pro/wata/sdk/model/Sorting.java
package pro.wata.sdk.model;

/**
 * Поле и направление сортировки для поисковых запросов, например
 * {@code orderId desc}. Направление по возрастанию не добавляет суффикса.
 */
public final class Sorting {

    private final String wireValue;

    private Sorting(String wireValue) {
        this.wireValue = wireValue;
    }

    public static Sorting asc(String field) {
        return new Sorting(field);
    }

    public static Sorting desc(String field) {
        return new Sorting(field + " desc");
    }

    /** Значение, отправляемое в query-параметре {@code Sorting}. */
    public String wireValue() {
        return wireValue;
    }

    @Override
    public String toString() {
        return wireValue;
    }
}

File: dotnet/src/Wata.Sdk/Models/Common/Sorting.cs
namespace Wata.Sdk.Models;

/// <summary>Поле сортировки для поиска ссылок и транзакций.</summary>
public enum SortField
{
    OrderId,
    CreationTime,
    Amount,
}

public enum SortDirection
{
    Ascending,
    Descending,
}

/// <summary>
/// Значение параметра Sorting. Сериализуется в вид "orderId" или "orderId desc",
/// как того требует API.
/// </summary>
public readonly record struct Sorting(SortField Field, SortDirection Direction = SortDirection.Ascending)
{
    public override string ToString()
    {
        var field = Field switch
        {
            SortField.OrderId => "orderId",
            SortField.CreationTime => "creationTime",
            SortField.Amount => "amount",
            _ => throw new ArgumentOutOfRangeException(nameof(Field), Field, null),
        };

        return Direction == SortDirection.Descending ? field + " desc" : field;
    }
}
